components/renku_data_services/solr/solr_client.py

```diff
@@ -364,3 +413,71 @@ async def delete(self, query: str) -> Response:
     async def close(self) -> None:
         """Close this client and free resources."""
         return await self.delegate.aclose()
+
+
+class SolrAdminClient(AbstractAsyncContextManager, ABC):
+    """A client to the core admin api.
+
+    Url: https://solr.apache.org/guide/solr/latest/configuration-guide/coreadmin-api.html
+    """
+
+    @abstractmethod
+    async def status(self, core_name: str | None) -> dict[str, Any] | None:
+        """Return the status of the connected core."""
+        ...
+
+    async def create(self, core_name: str | None) -> None:
+        """Create a core."""
+        ...
+
+
+class DefaultSolrAdminClient(SolrAdminClient):
+    """A client to the core admin api.
+
+    Url: https://solr.apache.org/guide/solr/latest/configuration-guide/coreadmin-api.html
+    """
+
+    delegate: AsyncClient
+    config: SolrClientConfig
+
+    def __init__(self, cfg: SolrClientConfig):
+        self.config = cfg
+        url_parsed = list(urlparse(cfg.base_url))
+        url_parsed[2] = urljoin(url_parsed[2], "/api/cores")
+        burl = urlunparse(url_parsed)
+        bauth = BasicAuth(username=cfg.user.username, password=cfg.user.password) if cfg.user is not None else None
+        self.delegate = AsyncClient(auth=bauth, base_url=burl, timeout=cfg.timeout)
+
+    async def __aenter__(self) -> Self:
+        await self.delegate.__aenter__()
+        return self
+
+    async def __aexit__(
+        self, exc_type: type[BaseException] | None, exc: BaseException | None, tb: TracebackType | None
+    ) -> None:
+        return await self.delegate.__aexit__(exc_type, exc, tb)
+
+    async def status(self, core_name: str | None) -> dict[str, Any] | None:
+        """Return the status of the connected core or the one given by `core_name`."""
+        core = core_name or self.config.core
+        resp = await self.delegate.get(f"/{core}")
+        if not resp.is_success:
+            raise SolrClientStatusException(self.config, resp)
+        else:
+            data = resp.raise_for_status().json()["status"][self.config.core]
```

Comment:
question: Do you want to use the new custom exceptions here that you defined above?

Reply:
Ah no thanks, just forgot (not used to marking it like this… 😅)